### ISSUE TITLE
🛠 Tooling: replace `jsonc-eslint-parser` with `@eslint/json`

### ISSUE THREAD

**Issue by rakleed** [JoshuaKGoldberg/eslint-plugin-package-json#655]
### Bug Report Checklist

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have pulled the latest `main` branch of the repository.
- [X] I have [searched for related issues](https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/issues?q=is%3Aissue) and found none that matched my issue.

### Overview

As suggested in https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/issues/601#issuecomment-2454704112, I am creating a separate issue to discuss replacing `jsonc-eslint-parser` dependency with `@eslint/json`.

[`@eslint/json`](https://bundlephobia.com/package/@eslint/json) is smaller and uses fewer dependencies than [`jsonc-eslint-parser`](https://bundlephobia.com/package/jsonc-eslint-parser). [The announcement post also says](https://eslint.org/blog/2024/10/eslint-json-markdown-support/#write-your-own-language-plugin) that it can be used as a parser for other plugins.

The main problem is that the plugin only works [since version 9.6.0 of ESLint](https://github.com/eslint/json#overview:~:text=Important%3A%20This%20plugin%20requires%20ESLint%20v9.6.0%20or%20higher%20and%20you%20must%20be%20using%20the%20new%20configuration%20system.), and also you can use only a flat config. So you will have to stop supporting eslint@8, but [now it still has more weekly downloads than eslint@9](https://www.npmjs.com/package/eslint?activeTab=versions).

**Comment by JoshuaKGoldberg** [comment]
Yeah 😞. As much as I'd love to use the new first-party package, I don't think it's time yet to drop support for ESLint 8. Once other popular plugins start to drop support then I think we can mark this as unblocked.

Thanks for filing!

**Comment by ota-meshi** [comment]
Hi there. I've created an experimental tool that converts `jsonc-eslint-parser` AST processing rules/plugins to be used with `@eslint/json`.
https://www.npmjs.com/package/eslint-json-compat-utils
Try it out if you want to use `eslint-plugin-package-json` and `@eslint/json` together now.


You can find a working [example online here].

[example online here]: https://eslint-online-playground.netlify.app/#eNqVVMtu2zAQ/BWCpzaIqKhJ0cK9FGhPObQFeixTgJVWDhOJFEgqqGHo37NL0or8SIL4IMuc2dnX0Fs+qPpercGXrbVl/iHuvDV8xbfSyCC5UT1IvmKSf+03ReYQX/LzxHgA57U1iXQhKnExQ752egieoC2LR3jYaYNPIoOndyaQP6MBfEbv4ltSIq2gXEaMbYB5cJgZq5WcKFOW4A08fIcBTAOm1pByz6lto/xtEvl7JapPoqoWKVI9Gf4oPj91MseKBv7Z0dR5Jscava1v1SzxYU/CbOqMVJfLOSE0OAhBg9vhoroSl7kzaSZ+nosTtTWtXmPbuCLdD9YFRgtjrbM9LSnRSjqT/Is0mZMXd72gJmYxdONam91mCwosUxJfOqht39Msm2fEfsXo1yWX4dtgv9l+UCEFTwfRRC8wLxKKMeguLdjA/xjcQKvGLrA/NJq42FZ34Fd4wM/O9kws+U2ebyoISTsnBMJ3w6fx75W6YvsVvjtq+H0KnaJEfOJ3WbLo5+vfP3+wWNVbauyUWY94HK1PS0hwRt0YBeb6M8fYohmHTtcqQHEPm2h3nKRzFq00NyiEWLQgothhBze4IGnQaMf/Awwlw2bInu9tg/FRnYDT943h59CNFNxhoflaJ8ryzj0DnnLES/ST9lsGED/equkRF9i8OQ==

**Comment by JoshuaKGoldberg** [comment]
That's fantastic, thanks @ota-meshi! Marking as accepting PRs to switch this package to using the compat utils.

Note that doing so is probably not a trivial change. It'll probably touch most or all of the rules in the plugin.

**Comment by JoshuaKGoldberg** [comment]
@all-contributors please add @rakleed for tool.

> 🤖 Beep boop! This comment was added automatically by [all-contributors-auto-action](https://github.com/marketplace/actions/all-contributors-auto-action).
> Not all contributions can be detected from Git & GitHub alone. Please comment any missing contribution types this bot missed.
> ...and of course, thank you for contributing! 💙

**Comment by michaelfaith** [comment]
> That's fantastic, thanks [@ota-meshi](https://github.com/ota-meshi)! Marking as accepting PRs to switch this package to using the compat utils.
> 
> Note that doing so is probably not a trivial change. It'll probably touch most or all of the rules in the plugin.

I'm just now catching up on this issue, since this discussion happened before my time on the project.  I'm not clear how the solution @ota-meshi provided unblocks this issue on our side.  It unblocks consuming projects from using our plugin with `@eslint/json`, with it's compat capability, but how would this unblock us from using the package?

**Comment by JoshuaKGoldberg** [comment]
> how would this unblock us from using the package within the plugin?

I was thinking we could author our rules using the plugin. That way they'll work both with the older `jsonc-eslint-parser` AST and with the newer `@eslint/json` one.

**Comment by azat-io** [comment]
Implementation is blocked by https://github.com/JoshuaKGoldberg/eslint-fix-utils/issues/55.

**Comment by michaelfaith** [comment]
https://github.com/JoshuaKGoldberg/eslint-fix-utils/issues/55 is no longer blocked, but I think we should keep this as blocked until we're ready to migrate fully to `@eslint/json` (which will require dropping support for eslint v8).

> I was thinking we could author our rules using the compat layer. That way they'll work both with the older `jsonc-eslint-parser` AST and with the newer `@eslint/json` one.

I may have a really poor imagination, haha, because I'm not sure how we could simultaneously support both, without providing two separate exports of the same rule (one that's piped through the compat function and one that isn't).  Is that what you mean?

**Comment by JoshuaKGoldberg** [comment]
Oh, maybe I was being to optimistic about support targets. In my mind there would be some logic that converts the AST style only if needed. As in:

* If the provided AST is the old one, the old style rules be given it directly
* If the provided AST is the new one, the compat layer would switch the AST to the old format, which would then be passed to rules

...or maybe vice versa?

Either way, I don't remember checking whether this existed or was something we'd have to build ourselves / contribute upstream.

**Comment by michaelfaith** [comment]
> Oh, maybe I was being to optimistic about support targets. In my mind there would be some logic that converts the AST style only if needed. As in:
> 
> * If the provided AST is the old one, the old style rules be given it directly
> * If the provided AST is the new one, the compat layer would switch the AST to the old format, which would then be passed to rules
> 
> ...or maybe vice versa?
> 
> Either way, I don't remember checking whether this existed or was something we'd have to build ourselves / contribute upstream.

The compat package wraps a rule, a plugin, or a create function that uses one parser, so that it's compatible with the other parser.  It doesn't have functions for converting one ast to another.  It's more intended for end users of plugins written for `jsonc-eslint-parser` to allow them to use that plugin with `@eslint/json`

**Comment by thernstig** [comment]
I found this when trying to use @eslint/json which by default does not work:

```ts
import json from '@eslint/json';
import { configs as packageJsonConfig } from 'eslint-plugin-package-json';
import { defineConfig, globalIgnores } from 'eslint/config';

export default defineConfig([
  globalIgnores(['.next/', 'next-env.d.ts']),
  // JSON linting
  {
    files: ['**/*.json'],
    ignores: ['package-lock.json'],
    plugins: { json },
    language: 'json/json',
    extends: ['json/recommended'],
  },
  // 'eslint-plugin-package-json' already contains a 'files' so no need here
  packageJsonConfig.recommended,
]);
```

The problem here is that the JSON block sets `language: 'json/json'` and it merges when run on eslint-plugin-package-json. It is afaic not fixable other than using `ignores: ['package-lock.json', 'package.json']`, but that also means the normal JSON linting is not done on package.json anymore.

**Comment by michaelfaith** [comment]
Yeah, there's not really going to be a good way to have `package.json`  files use two different languages.  Until our plugin adopts the `json/json` language, it's technically `language: "@/js"`.  So, you could add that explicitly to the packageJson config. 

```ts
{
  extends: [ packageJsonConfig.recommended ],
  language: "@/js",
}
```

but then you'd have the reverse problem of the other plugin trying to use the javascript language for its rules.

**Comment by anomiex** [comment]
The workaround package mentioned in https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/issues/655#issuecomment-2506852630 seems to work around it. We've been using it something like this:

```js
import eslintPluginPackageJson from 'eslint-plugin-package-json';
// @todo Remove use of eslint-json-compat-utils (and jsonc-eslint-parser) once https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/issues/655 is fixed.
import { toCompatPlugin as jsonToCompatPlugin } from 'eslint-json-compat-utils';
```
and then a config block like this
```js
		{
			name: 'Package.json',
			files: [ '**/package.json' ],
			plugins: {
				'package-json': jsonToCompatPlugin( eslintPluginPackageJson ),
			},
			rules: {
				...eslintPluginPackageJson.configs.recommended.rules,

				// Empty browserslist does something.
				'package-json/no-empty-fields': [ 'error', { ignoreProperties: [ 'browserslist' ] } ],
			},
		},
```

**Comment by michaelfaith** [comment]
Alright, I've added all of the sub-issues that I think will comprise our migration (except for a placeholder for all of the individual rule migration issues that I'll create closer to when we do the work).  But these all track against different parts of the migration plan defined in #1612 -> #1847 (https://github.com/michaelfaith/eslint-plugin-package-json/blob/main/docs/plans/parser-to-language-migration/plan.md)

**Comment by michaelfaith** [comment]
Woohoo!  Happy to share that our latest release (1.2.0) includes a new `experimental` entry point, with versions of the `rules`, `configs`, and `plugin` that work with `@eslint/json` instead of `jsonc-eslint-parser`. 🎉   You can find out more on our Getting Started page: https://eslint-plugin-package-json.dev/getting-started/#-experimental-api.  This marks completion of Milestone 1 in our [migration plan](https://github.com/michaelfaith/eslint-plugin-package-json/blob/main/docs/plans/parser-to-language-migration/plan.md).  In our next major version, we'll replace the primary entry point with the experimental one (Milestone 2). 

Please give it a spin and see how it works for you.  If you find any issues with it, please raise them as new issues.  This issue will remain open to track the remainder of the migration.  Appreciate your patience as we worked through our plan for this.

**Comment by michaelfaith** [comment]
Currently the following repos that I work in are dogfooding the `experimental` api:
- https://github.com/michaelfaith/package-json-validator
- https://github.com/michaelfaith/eslint-fix-utils
- https://github.com/flint-fyi/flint

If you're testing it out and have any feedback to share (either positive or negative), i'd love to know.

**Comment by anomiex** [comment]
We've been using it in https://github.com/Automattic/jetpack/ for about a month now (and before that we were using `eslint-json-compat-utils` to make it work). No problems found.

**Comment by michaelfaith** [comment]
Awesome! Great to hear 👌
